### DISCUSSION TITLE
Gracefully reconnect without losing messages

### DISCUSSION THREAD

**Discussion by intellix** [enisdenjo/graphql-ws#603] [Q&A]
We're using graphql-ws in the server which are running within K8S Pods that scale up/down. When K8S scales up/down it'll move traffic between different pods and we're seeing an issue with losing important messages in the process. 

Our idea is that when a SIGTERM is received on the server, we'll emit a message through the socket to tell the client to reconnect, but in order to not miss messages, it would be great if we could blue/green replace the socket connecting to another pod. 

I believe in order to do that, we'd need to momentarily open up a secondary socket before closing the first one (which is scheduled to be closed within seconds). 

Just wondering if you have any recommendation for achieving this? I'm thinking instead of having an ApolloLink containing 1x client, we would have an array of clients and upon receiving a SIGKILL message from a socket, we would create a 2nd client.

The Observable returned by request would then listen to both clients and dedupe duplicate emits. 

Thoughts? Am I running into a trap somewhere?

**Comment by enisdenjo** [comment]
That sounds like a completely viable solution. I don't think you'll run into any issues aside from making sure the deduplication works as expected. 

An alternative approach would be an implementation outside of `graphql-ws`, inside the actual subscription iterable in `graphql` itself. What you can do is have a pubsub singleton that is emitting the subscription events and `graphql` from `graphql-ws` subscribing to that pubsub using a unique identifier from the context. This way, the pubsub can have a "debounce" before the actual subscription is completed, during the "debounced" period (when there are no subscribers but the timeout didnt expire), the pubsub would queue up the events and flush them once a `graphql` subscribes matching the identifier from the context. I hope I conveyed the idea. 😅  If not, feel free to ask questions - I'd be happy to help out where I can.

**Reply by enisdenjo** [reply]
Of course, you have to be careful because the new subscription might not happen and those queued events need to be destroyed after the debounce timeout expires.
